frame/evm/precompile/contracts/kton/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+//
+// Copyright (C) 2018-2021 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::traits::{Currency, ExistenceRequirement};
+use sha3::Digest;
+use sp_core::{H160, U256};
+use sp_runtime::traits::UniqueSaturatedInto;
+use sp_std::marker::PhantomData;
+use sp_std::prelude::*;
+use sp_std::vec::Vec;
+
+use core::str::FromStr;
+use darwinia_evm::{AddressMapping, Runner, Trait};
+use darwinia_evm_primitives::Precompile;
+use ethabi::{Function, Param, ParamType, Token};
+use evm::{Context, ExitError, ExitSucceed};
+use sp_std::borrow::ToOwned;
+
+const DEPOSIT_FUNC: &[u8] = b"deposit(address,uint256)";
+const KTON_PRECOMPILE: &str = "0000000000000000000000000000000000000016";
+/// Kton Precompile Contract
+///
+/// The contract address: 0000000000000000000000000000000000000016
+pub struct Kton<T: Trait> {
+	_maker: PhantomData<T>,
+}
+
+impl<T: Trait> Precompile for Kton<T> {
+	// The execute process of Kton precompile, the input data consists of four parts:
+	// 1. Wkton solidity contract deployed address(20 bytes)
+	// 2. The function selector(4 bytes)
+	// 3. The deposit kton evm address(20 bytes)
+	// 4. The deposit kton value(32 bytes)
+	fn execute(
+		input: &[u8],
+		target_limit: Option<usize>,
+		context: &Context,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError> {
+		let helper = U256::from(10)
+			.checked_pow(U256::from(9))
+			.unwrap_or(U256::MAX);
+		let deposit_func = &sha3::Keccak256::digest(&DEPOSIT_FUNC)[0..4];
+		// decode input
+		let input = InputData::decode(input)?;
+
+		let con_caller = T::AddressMapping::into_account_id(context.caller);
+		if hex::encode(input.func_selector) == hex::encode(&deposit_func) {
+			// 1. Transfer kton from sender to kton erc20 contract
+			let wkton_account_id = T::AddressMapping::into_account_id(input.wkton_address);
+			let transfer_value = input.value.saturating_mul(helper);
+			T::KtonCurrency::transfer(
+				&con_caller,
+				&wkton_account_id,
+				transfer_value.low_u128().unique_saturated_into(),
+				ExistenceRequirement::AllowDeath,
+			)
+			.map_err(|_| ExitError::Other("Transfer in Kton precompile failed".into()))?;
+
+			// 2. Call wkton sol contract deposit
+			let raw_input = make_deposit_input(input.address, input.value)?;
```

Comment:
e.g. calldata is `selector("deposit(from, value)")`

```suggestion
			let raw_input = make_deposit_input(calldata, input.address, input.value)?; 
```